runtime/jcl/common/sun_misc_Unsafe.cpp

```diff
@@ -96,19 +95,66 @@ Java_sun_misc_Unsafe_defineAnonymousClass(JNIEnv *env, jobject receiver, jclass
 		hostClassLoader = vm->systemClassLoader->classLoaderObject;
 	}
 	jobject hostClassLoaderLocalRef = vmFuncs->j9jni_createLocalRef(env, hostClassLoader);
+
+	J9ClassPatchMap cpPatchMap = {0, NULL};
+	j9array_t patchArray = NULL;
+	PORT_ACCESS_FROM_ENV(env);
+
+	U_16 indexMap[BUFFER_SIZE];
+	if (constPatches != NULL) {
+		patchArray = (j9array_t)J9_JNI_UNWRAP_REFERENCE(constPatches);
+		cpPatchMap.size = (U_16)J9INDEXABLEOBJECT_SIZE(currentThread, patchArray);
+		if (cpPatchMap.size <= BUFFER_SIZE) {
+			cpPatchMap.indexMap = indexMap;
+		} else {
+			cpPatchMap.indexMap = (U_16 *)j9mem_allocate_memory(cpPatchMap.size * sizeof(U_16), J9MEM_CATEGORY_VM);
+		}
+	}
+
 	vmFuncs->internalExitVMToJNI(currentThread);
 
 	jsize length = env->GetArrayLength(bytecodes);
 
 	/* acquires access internally */
-	jclass anonClass = defineClassCommon(env, hostClassLoaderLocalRef, NULL,bytecodes, 0, length, protectionDomainLocalRef, J9_FINDCLASS_FLAG_UNSAFE | J9_FINDCLASS_FLAG_ANON, hostClazz);
+	jclass anonClass = defineClassCommon(env, hostClassLoaderLocalRef, NULL,bytecodes, 0, length, protectionDomainLocalRef, J9_FINDCLASS_FLAG_UNSAFE | J9_FINDCLASS_FLAG_ANON, hostClazz, &cpPatchMap);
 	if (env->ExceptionCheck()) {
 		return NULL;
 	} else if (NULL == anonClass) {
 		throwNewInternalError(env, NULL);
 		return NULL;
 	}
 
+	if (constPatches != NULL) {
+		vmFuncs->internalEnterVMFromJNI(currentThread);
+		J9Class *clazz = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, J9_JNI_UNWRAP_REFERENCE(anonClass));
+
+		j9object_t item = NULL;
+		U_32 *cpShapeDescription = J9ROMCLASS_CPSHAPEDESCRIPTION(clazz->romClass);
+		J9ConstantPool *ramCP = J9_CP_FROM_CLASS(clazz);
+		J9ROMConstantPoolItem *romCP = ramCP->romConstantPool;
+
+		for (U_16 i = 0; i < cpPatchMap.size; i++) {
+			item = J9JAVAARRAYOFOBJECT_LOAD(currentThread, patchArray, i);
+			if ((item != NULL) && (J9_CP_TYPE(cpShapeDescription, cpPatchMap.indexMap[i]) == J9CPTYPE_STRING)) {
+				J9UTF8 *romString = J9ROMSTRINGREF_UTF8DATA((J9ROMStringRef *)&romCP[cpPatchMap.indexMap[i]]);
+
+				for (U_16 j = 1; j < clazz->romClass->ramConstantPoolCount; j++) {
+					if ((J9_CP_TYPE(cpShapeDescription, j) == J9CPTYPE_STRING)
+						&& J9UTF8_EQUALS(romString, J9ROMSTRINGREF_UTF8DATA((J9ROMStringRef *)&romCP[j]))
+					) {
+						J9RAMStringRef *ramStringRef = ((J9RAMStringRef *)ramCP) + j;
+						J9STATIC_OBJECT_STORE(currentThread, clazz, &ramStringRef->stringObject, item);
+					}
+				}
+			}
+		}
+
+		if (cpPatchMap.size > BUFFER_SIZE) {
+			j9mem_free_memory(cpPatchMap.indexMap);
```

Comment:
Good practice to immediately `cpPatchMap.indexMap = NULL` here

Reply:
These are @fengxue-IS changes for Constant Pool Patching (#8460). These should go away after #8460 is merged, and this pull request is rebased. Only the seven commits listed in the header description need to be reviewed for this pull request.